### PR TITLE
#29 + #30 winningest_coach, worst_coach and some unsolicited strong opinions

### LINKED ISSUE #29: Method: winningest_coach
**Goals**
Create a method in the `StatTracker` class that returns the came of the Coach with the best win percentage for the season.

Takes a `season_id` as an argument and returns a `string`.

```ruby
stat_tracker.is_a?(StatTracker) # => true

stat_tracker.winningest_coach # => "Mike Malone"
```

**Ideas** 

**Additional context**
* [Link from project requirements](https://backend.turing.edu/module1/projects/futbol_pd/iterations/file_io_stats)
* very similar to `worst_coach`

### LINKED ISSUE #30: Method: worst_coach
**Goals**
Create a method in the `StatTracker` class that returns the came of the Coach with the worst win percentage for the season.

Takes a `season_id` as an argument and returns a `string`.

```ruby
stat_tracker.is_a?(StatTracker) # => true

stat_tracker.worst_coach # => "Mark Jackson"
```

**Ideas** 

**Additional context**
* [Link from project requirements](https://backend.turing.edu/module1/projects/futbol_pd/iterations/file_io_stats)
* very similar to `winningest_coach`

### PR DESCRIPTION
**Pull request by Daniel-Gallagher92**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] I have read the **CONTRIBUTING** document.
- [] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

closes #29 
closes #30

### CODE REVIEW

**Review by HaydenMcSorley — APPROVED** [review]
Tortorella is a chump

**Review by ConnorRichmond — APPROVED** [review]
looks good, tests clear

File: lib/stat_tracker.rb [hunk 1]
```diff
@@ -230,4 +230,58 @@ def fewest_tackles(season)
...
+    end
+
+    game_ids = games_for_season.map do |game| 
+      game.game_id
+    end
+
+    @game_teams.find_all do |game_team|
+      game_ids.include?(game_team.game_id)
+    end
+  end
```
Comment (ethanrossblack) [inline lib/stat_tracker.rb:247]: I made a very similar helper method!

File: lib/stat_tracker.rb [hunk 2]
```diff
@@ -230,4 +230,58 @@ def fewest_tackles(season)
   
     fewest_tackles_team&.team_name
   end
+
+  def season_games(season)
```
Comment (ethanrossblack) [inline lib/stat_tracker.rb:234]: Maybe rename this season_game_teams because it's collecting GameTeams?

Comment (ethanrossblack) [reply lib/stat_tracker.rb:234]:
```suggestion
  def season_game_teams(season) 
```

File: lib/stat_tracker.rb [hunk 3]
```diff
@@ -230,4 +230,58 @@ def fewest_tackles(season)
...
+    end
+
+    @game_teams.find_all do |game_team|
+      game_ids.include?(game_team.game_id)
+    end
+  end
+
+  def season_wins(season) 
+    # Returns a hash of head coaches and their win count for a given season
+    season_games = season_games(season)
```
Comment (ethanrossblack) [inline lib/stat_tracker.rb:251]:
```suggestion
    season_game_teams = season_game_teams(season)
```

File: lib/stat_tracker.rb [hunk 4]
```diff
@@ -230,4 +230,58 @@ def fewest_tackles(season)
...
+    head_coaches = season_games.group_by do |game| 
+      game.head_coach
```
Comment (ethanrossblack) [inline lib/stat_tracker.rb:253]:
```suggestion
    head_coaches = season_game_teams.group_by do |game_team| 
      game_team.head_coach
```

File: lib/stat_tracker.rb [hunk 5]
```diff
@@ -230,4 +230,58 @@ def fewest_tackles(season)
...
+    end
+
+    head_coaches.each do |coach, games|
+      head_coaches[coach] = games.find_all do |game| 
+        game.result == "WIN"
+      end.count
```
Comment (ethanrossblack) [inline lib/stat_tracker.rb:259]:
```suggestion
    head_coaches.each do |coach, game_teams|
      head_coaches[coach] = game_teams.find_all do |game_teams| 
        game_team.result == "WIN"
      end.count
```

**Review by ethanrossblack — CHANGES_REQUESTED** [review]
Change some variable names to clarify the use of `game_teams` vs `games`

**Review by ethanrossblack — APPROVED** [review]

**Pull request merged**